Online connectivity check in connection manager (#63531)
zephyrproject-rtos/zephyr: samples/net/connectivity_check/README.rst

   :relevant-api: bsd_sockets http_client tls_credentials secure_sockets_options

   Use online connectivity checker in connection manager to verify that
   we have connectivity to a user-specified network/host.

Overview
********

This sample application enables the online connectivity checker in network
connection manager. The checker tries to connect to user-specified host either
by an ICMP ping request, or by a HTTP(S) GET request.
If the connection can be established, a ``NET_EVENT_CONNECTIVITY_ONLINE`` is
generated and application can monitor this event and then act on it.
If the connectivity to user-specified host is lost, then
a ``NET_EVENT_CONNECTIVITY_OFFLINE`` event is generated.

The source code for this sample application can be found at:
:zephyr_file:`samples/net/connectivity_check`.

Requirements
************

- :ref:`networking_with_host`

Building and Running
********************

You can use this application on a supported board, including
running it inside QEMU as described in :ref:`networking_with_qemu`.

Build the http-client sample application like this:

.. zephyr-app-commands::
   :zephyr-app: samples/net/connectivity_check

>    :board: <board to use>

Author: Hmm, I don't think the instructions are very qemu centric. I mention qemu once in the readme file. Perhaps we could call this emulator centric instructions as also native_sim board is mentioned.

>    :board: <board to use>

Author: I added more information if user wants to use some other board to run the sample.

>    :board: <board to use>

Reviewer: Ya, maybe emulator centric then :) My concern is with the "As the connectivity check is meant to connect to some outside network with public IP addresses, it might be necessary to setup network settings on the host side so that Linux host can act as a router." paragraph -- when I first read it I was literally like "wait, what, it said earlier I can run this sample on a supported board, so what is this referring to?"

Maybe turn the two following paragraphs into a `note` aimed at people running in an emulated environment?

>    :board: <board to use>

Author: I changed the chapter to `Running connectivity_check against Docker Container`, would that be enough here? Also added information that these instructions are meant for testing purposes.

>    :board: <board to use>

Reviewer: please push your changes tho :)

on the host side so that Linux host can act as a router.

It is possible to see how the checker works by utilizing a Docker container
from the `net-tools <https://github.com/zephyrproject-rtos/net-tools>`_ project.

Running connectivity-checker in Docker Container
================================================

You can run this ``connectivity-check`` sample application in Linux host and
make it connect to a Docker container which contains needed server components
like HTTP and HTTPS servers, a DNS server and DHCP server.

Get the ``net-tools`` project and build the net-tools container.

In a terminal window:

.. code-block:: console

    $ cd tools/net-tools
    $ cd docker
    $ docker build -t net-tools .

Run the test script which compiles the connectivity-check application for ``native-sim``
board, then starts the Docker container, and finally starts the ``zephyr.exe``
application process which uses the network services provided by the container.

.. code-block:: console

    $ cd $ZEPHYR_BASE
    $ scripts/net/run-sample-tests.sh samples/net/connectivity_check

Note that ``run-sample-tests.sh`` script can be stopped manually by pressing

> :kbd:`CTRL+A` :kbd:`x`. The application will run couple of minutes and tries to
> verify the online connectivity.

Reviewer: ```suggestion
:kbd:`CTRL+A` :kbd:`x`. The application will run for a couple of minutes and try to
verify the online connectivity.
```

.. zephyr:code-sample:: online-connectivity-check
   :name: Online Connectivity Checker for Connection Manager
   :relevant-api: bsd_sockets http_client tls_credentials secure_sockets_options

   Use online connectivity checker in connection manager to verify that
   we have a connectivity to a user specified network/host.

Overview
********

This sample application enables the online connectivity checker in network

> connection manager. The checker tries to connect to user specific host either

Reviewer: ```suggestion
connection manager. The checker tries to connect to user-specified host either
```

> by a ICMP ping request, or by a HTTP(S) GET request.

Reviewer: ```suggestion
by an ICMP ping request, or by a HTTP(S) GET request.
```

If the connection can be established, a ``NET_EVENT_CONNECTIVITY_ONLINE`` is
generated and application can monitor this event and then act on it.

> If the connectivity to user specified host is lost, then

Reviewer: ```suggestion
If the connectivity to user-specified host is lost, then
```

a ``NET_EVENT_CONNECTIVITY_OFFLINE`` event is generated.

The source code for this sample application can be found at:
:zephyr_file:`samples/net/connectivity_check`.

Requirements
************

- :ref:`networking_with_host`

Building and Running
********************

You can use this application on a supported board, including
running it inside QEMU as described in :ref:`networking_with_qemu`.

Build the http-client sample application like this:

.. zephyr-app-commands::
   :zephyr-app: samples/net/connectivity_check
   :board: <board to use>
   :conf: <config file to use>
   :goals: build
   :compact:

As the connectivity check is meant to connect to some outside network
with public IP addresses, it might be necessary to setup network settings

> in host side so that Linux host can act as a router.

Reviewer: ```suggestion
on the host side so that Linux host can act as a router.
```
